# Проект по SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.
Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

### Задачи, необходимые запросы: 

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [2]:
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
db_config['pwd'],
db_config['host'],
db_config['port'],
db_config['db'])
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [3]:
def select(sql):
    return pd.io.sql.read_sql(sql, con = engine)

In [4]:
query = '''
SELECT * FROM books
'''
books = pd.io.sql.read_sql(query, con = engine)
query_1 = '''
SELECT * FROM authors
'''

authors = pd.io.sql.read_sql(query_1, con = engine)
query_2 = '''
SELECT * FROM publishers
'''

publishers = pd.io.sql.read_sql(query_2, con = engine)
query_3 = '''
SELECT * FROM ratings
'''

ratings = pd.io.sql.read_sql(query_3, con = engine)
query_4 = '''
SELECT * FROM reviews
'''

reviews = pd.io.sql.read_sql(query_4, con = engine)

#### Таблица books

In [10]:
query_1 = '''
         SELECT *
         FROM books
         LIMIT 5
        '''
books = pd.io.sql.read_sql(query_1, con = engine)
books

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

#### Таблица authors

In [11]:
query_2 = '''
         SELECT *
         FROM authors
         LIMIT 5
        '''
authors = pd.io.sql.read_sql(query_2, con = engine)
authors

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

#### Таблица publishers

In [12]:
query_3 = '''
         SELECT *
         FROM publishers
         LIMIT 5
        '''
publishers = pd.io.sql.read_sql(query_3, con = engine)
publishers

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

#### Таблица ratings

In [13]:
query_4 = '''
         SELECT *
         FROM ratings
         LIMIT 5
        '''
ratings = pd.io.sql.read_sql(query_4, con = engine)
ratings

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

#### Таблица reviews

In [14]:
query_5 = '''
         SELECT *
         FROM reviews
         LIMIT 5
        '''
reviews = pd.io.sql.read_sql(query_5, con = engine)
reviews

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

### Книги, выпущенные после 1 января 2000 года

In [5]:
books_cnt = '''
SELECT COUNT(book_id) AS books_cnt FROM books WHERE publication_date >= '2000-01-02'
'''
select(books_cnt)

books_cnt
0        819

С 1 января 2000 года было выпущено 819 книг.

### Количество обзоров и средняя оценка

In [16]:
ratings_reviews = '''
SELECT 
    books.title AS title,
    books.book_id AS id,
    AVG(ratings.rating) AS avg_rating,
    COUNT(DISTINCT reviews.text) AS cnt_reviews
FROM
    books
INNER JOIN ratings ON ratings.book_id = books.book_id
INNER JOIN reviews ON reviews.book_id = books.book_id
GROUP BY
     books.title,
     books.book_id
order by
     cnt_reviews DESC
limit 20
'''
select(ratings_reviews)

title   id  avg_rating  \
0                             Twilight (Twilight  #1)  948    3.662500   
1                                       The Alchemist  627    3.789474   
2                           Outlander (Outlander  #1)  497    4.125000   
3                 The Hobbit  or There and Back Again  750    4.125000   
4                                     Eat  Pray  Love  207    3.395833   
5                                      The Book Thief  656    4.264151   
6   The Curious Incident of the Dog in the Night-Time  695    4.081081   
7                              The Catcher in the Rye  673    3.825581   
8                           The Giver (The Giver  #1)  733    3.750000   
9                                    The Glass Castle  734    4.206897   
10  The Lightning Thief (Percy Jackson and the Oly...  779    4.080645   
11                                           The Road  854    3.772727   
12                                Water for Elephants  963    3.977273   
13             The Da Vinci Code (Robert Langdon  #2)  696    3.830508   
14  Harry Potter and the Chamber of Secrets (Harry...  299    4.287500   
15  Harry Potter and the Prisoner of Azkaban (Harr...  302    4.414634   
16     A Clash of Kings  (A Song of Ice and Fire  #2)   15    4.391304   
17                Extremely Loud and Incredibly Close  232    3.416667   
18  Freakonomics: A Rogue Economist Explores the H...  267    3.909091   
19                                            Dracula  194    3.296296   

    cnt_reviews  
0             7  
1             6  
2             6  
3             6  
4             6  
5             6  
6             6  
7             6  
8             6  
9             6  
10            6  
11            6  
12            6  
13            6  
14            6  
15            6  
16            5  
17            5  
18            5  
19            5

Для 7 книг отсуствуют рейтинг и обзор, максимальное кол-во обзоров у книги Twilight (Twilight #1). Достаточно много книг с оценкой 5, но у них всего 2 обзора.

### Издательство, которое выпустило наибольшее число книг толще 50 страниц. Исключим брошюры

In [ ]:
query_8 = '''
         SELECT publishers.publisher_id, publishers.publisher, COUNT(books.book_id) as number_of_books
         FROM publishers
         JOIN books ON publishers.publisher_id = books.publisher_id
         WHERE books.num_pages > 50
         GROUP BY publishers.publisher_id
         ORDER BY number_of_books DESC
         LIMIT 1
        '''
pd.io.sql.read_sql(query_8, con = engine)

Penguin Books (publisher_id = 212) выпустило 42 книги, в которых больше 50 страниц.

### Автор с самой высокой средней оценкой книги — только книги с 50 и более оценками

In [ ]:
query_9 = '''
         SELECT authors.author_id, authors.author, AVG(avg_rating) AS avg_rating_total
         FROM authors
         JOIN books ON books.author_id = authors.author_id
         JOIN(
             SELECT books.book_id, books.title, COUNT(rating_id) AS number_of_ratings, AVG (rating) AS avg_rating
             FROM books
             JOIN ratings ON ratings.book_id = books.book_id
             GROUP BY books.book_id, books.title) AS subq ON subq.book_id = books.book_id
         WHERE number_of_ratings >= 50
         GROUP BY authors.author_id, authors.author
         ORDER BY avg_rating_total DESC

        '''
pd.io.sql.read_sql(query_9, con = engine)

Автор с самой высокой средней оценкой книг (среди авторов, у которых 50 и более оценок) - J.K. Rowling/Mary GrandPré (author_id = 236), ее средняя оценка книг - 4.28.

### Среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [ ]:
#Найдем среднее число ревью, написанных пользователями, которые оставили больше 50 оценок
query_13 = '''
     SELECT AVG(count)
     FROM (
         SELECT reviews.username, COUNT(reviews.review_id)
         FROM reviews
         JOIN (
                 SELECT *
         FROM
         (SELECT ratings.username, COUNT(ratings.rating_id) as number_of_ratings
         FROM ratings
         GROUP BY ratings.username
         ORDER BY number_of_ratings DESC
         ) AS subq
         WHERE number_of_ratings > 50
         ) AS subq_2 ON subq_2.username = reviews.username
         GROUP BY reviews.username
         ) AS subq_3
     '''
pd.io.sql.read_sql(query_13, con = engine)

Среднее кол-во текстовых обзоров пользователей, которые поставили более чем по 50 оценок равно 24.33

<div class="alert alert-block alert-success">
<b>Комментарии от ревьюера:</b>

И это .. правильный ответ! Ты корректно отобрал пользователей и рассчитал их показатели.

В итоге, 4\5 с первой попытки - это достойный результат)
</div>